jax/lax/lax_control_flow.py

```diff
@@ -474,10 +476,16 @@ def scan(f, init, xs):
   x_dtypes = [x.dtype for x in xs_flat]
   x_avals = tuple(_map(ShapedArray, x_shapes, x_dtypes))
   jaxpr, consts, out_tree = _initial_style_jaxpr(f, in_tree, carry_avals + x_avals)
-  carry_avals_out, y_avals = split_list(jaxpr.out_avals, [num_carry])
-  if tuple(carry_avals_out) != carry_avals:
+
+  out_tree_avals = tree_unflatten(out_tree, jaxpr.out_avals)
+  init_tree_avals = tree_unflatten(init_tree, carry_avals)
+  if not isinstance(out_tree_avals, tuple) or len(out_tree_avals) != 2:
+    msg = "scan body output must be a pair, got {}."
+    raise TypeError(msg.format(out_tree_avals))
+  if out_tree_avals[0] != init_tree_avals:
```

Comment:
While our standard pytrees (`list`/`tuple`/`dict`) define equality, this line would also require user-defined pytree types to define equality. That may be a new constraint on pytrees. I'm fine with adding that constraint here, but an alternative would be to compare the treedefs and separately compare the flattened lists of avals.

Reply:
Good point, I had tried that but I was not able to extract the treedef for the first element of a pair given the treedef for the pair. I need this to check that the returned carry (first element) matches the initial carry.

Reply:
I haven't looked at the details but [`treedef_children`](https://github.com/google/jax/blob/5b724e0019243cc5ebb71d65963e1d829f65999e/jax/tree_util.py#L122) might help us here.

Reply:
Implemented suggestion. Custom pytrees can still be tricky, if the custom flattener stores aux_data that does not support equality.